pvl-tools/pvl.cson

```diff
@@ -30,67 +30,104 @@ services:
       , error: ["NOT_FOUND", "matching DNS entry not found"] } },
   ]]
   facebook: [[
-    # check url and extract username
+    # Check that the claimed username has no slashes or funny business that
+    # might trick us into checking a different url.
+    # Facebook usernames don't actually allow any special characters, but it's
+    # still possible for a malicious user to *claim* they have some slashes
+    # and a question mark in their name, in the hopes that that will trick us
+    # into hitting a totally unrelated URL. Guard against that happening by
+    # checking for special characters in the claimed name.
+    { assert_regex_match: {
+      , pattern: "^[a-zA-Z0-9\\.]+$"
+      , from: "username_service"
+      , error: ["BAD_USERNAME", "Invalid characters in username '%{username_service}'"] } },
+
+    # Check the provided url and extract the username and path.
+    # Accept either mobile or desktop urls. The fetched url will be rewritten later.
+    # We want to be strict about the structure of the url.
+    # No query parameters, no unexpected characters in the post ID.
     { regex_capture: {
-      , pattern: "^https://m\\.facebook\\.com/([^/]*)/posts/.*$"
+      , pattern: "^https://(m|www)\\.facebook\\.com/([^/]*)/posts/([0-9]+)$"
       , from: "hint_url"
-      , into: ["username_from_url"]
-      , error: ["BAD_API_URL", "Bad hint from server; URL should start with 'https://m.facebook.com/%{username_service}/posts/', received '%{hint_url}'"] } },
+      , into: ["unused1", "username_from_url", "post_id"]
+      , error: ["BAD_API_URL", "Bad hint from server; URL should start with 'https://m.facebook.com/%{username_service}/posts/', got '%{hint_url}'"] } },
+
+    # Check that the claimed username matches the url.
+    # Checking for the correct username is essential here. We rely on this
+    # check to prove that the user in question actually wrote the post. (Note
+    # that the m-site does *not* enforce this part of the URL. Only the
+    # desktop site does.)
     { assert_compare: {
       , cmp: "stripdots-then-cicmp"
```

Comment:
Now that we have ReplaceAll, could we just replace all dots with the empty string and then do a regular comparison? Maybe we can kill this special comparison type? No rush though.